### PR TITLE
Libraries added for OLED Display

### PR DESCRIPTION
**Pull request by ptal78**
These two libraries gfx and tank help to draw lines and other elements on the OLED display for the ESP32

Also, the ssd1309.py file is the same as the previous version SSD1309.py, but importantly capital letters don't seem to be compatible with other people's code. You might like to simply adjust the existing file name, or ignore this part.

### CODE REVIEW

**Comment by ptal78** [comment]
Sorry. I seem to have committed code that I shouldn't have. I'll try to undo this.

**Comment by ptal78** [comment]
I got a shock when the offline version file was automatically updated - LOL.

**Comment by ptal78** [comment]
Have I edited these files correctly please? I'm not sure how to test them out on my fork yet.

File: ui/core/block_definitions.js [hunk 1]
```diff
@@ -986,6 +998,59 @@ Blockly.Blocks['write_oled'] = {
   }
 };
 
+Blockly.Blocks['tank_init'] = {
```
Comment (JorgeGMarques) [inline ui/core/block_definitions.js:1001]: `init_tank` instead of `tank_init`

File: ui/toolbox/esp32.xml [hunk 2]
```diff
@@ -697,7 +697,9 @@
 
     <category name="OLED Display">
       <label text="SSD1306 OLED I2C Display"></label>
-      <button text="Install SSD1306 library" callbackKey="installPyLib"></button>
+      <button text="Install ssd1306 library" callbackKey="installPyLib"></button>
```
Comment (JorgeGMarques) [inline ui/toolbox/esp32.xml:700]: We already have this library as `SSD1306`

File: ui/core/generator_stubs.js [hunk 3]
```diff
@@ -397,6 +402,37 @@ Blockly.Python['write_oled'] = function(block) {
...
+
+  return code;
+};
+
+Blockly.JavaScript['tank_move'] = function(block) {
+  var dropdown_option = block.getFieldValue('option');
+  var value_move = Blockly.JavaScript.valueToCode(block, 'Move', Blockly.JavaScript.ORDER_ATOMIC);
+  // TODO: Assemble JavaScript into code variable.
+  var code = 'tank1.move(' + value_move + ', 1)\n';	
+  if (dropdown_option === 'Prime') {
```
Comment (JorgeGMarques) [inline ui/core/generator_stubs.js:422]: Here is
`  if (dropdown_option === ' TankPrime') {`

File: ui/core/generator_stubs.js [hunk 4]
```diff
@@ -397,6 +402,37 @@ Blockly.Python['write_oled'] = function(block) {
...
+  	code = 'tank1.move(' + value_move + ', 2)\n';
+  } else if (dropdown_option === 'Jump') {
```
Comment (JorgeGMarques) [inline ui/core/generator_stubs.js:424]: Here is
`  if (dropdown_option === ' TankJump') {`

File: ui/pylibs/ssd1306.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
+#MicroPython SSD1306 OLED driver, I2C and SPI interfaces created by Adafruit
```
Comment (JorgeGMarques) [inline ui/pylibs/ssd1306.py:1]: We already have this library as SSD1306

**Review by JorgeGMarques — CHANGES_REQUESTED** [review]
Hi! great work, I just  reviewed it.
About the `index_offline.html`, it's the bot that should be disabled...
Can you disable it in your fork at https://github.com/ptal78/BIPES_ui_testing/actions ?
Screenshot from 2021-08-22 09-35-07

Thanks! after that I will just delete `index_offline.html` and merge!

File: ui/pylibs/tank.py [hunk 6]
```diff
@@ -0,0 +1 @@
+
```
Comment (JorgeGMarques) [inline ui/pylibs/tank.py:1]: It's blank...

**Comment by ptal78** [comment]
> Hi! great work, I just reviewed it.
> About the `index_offline.html`, it's the bot that should be disabled...
> Can you disable it in your fork at https://github.com/ptal78/BIPES_ui_testing/actions ? Just click the 3 dots and choose "disable workflow".
> Screenshot from 2021-08-22 09-35-07
> 
> Thanks! after that I will just delete `index_offline.html` and merge!

I've disabled that now.

**Comment by ptal78** [comment]
I think the xml file is wrong as it doesn't match the js of the block.

**Comment by JorgeGMarques** [comment]
Fixed some stuff, please `git pull` there to apply changes.
then include the code for `ui/pylibs/tank.py ` (it's empty).

File: ui/toolbox/esp32.xml (hunk 2 above)
Comment (ptal78) [reply ui/toolbox/esp32.xml]: When it is installed with capital letters 'SSD' it doesn't seem to work. The lowercase version 'ssd' is seemingly the library everywhere else. I suggest lowercase could be used to be compatible with existing uses.

**Comment by ptal78** [comment]
Hi Jorge, I'm not sure what has happened, but the OLED menu has now put all the blocks on top of each other.
Are you able to fix that?
Phil

**Comment by ptal78** [comment]
Works great now. Cheers

**Pull request merged**